src/libbson/src/jsonsl/jsonsl.c

```diff
@@ -457,13 +457,17 @@ jsonsl_feed(jsonsl_t jsn, const jsonsl_char_t *bytes, size_t nbytes)
                 } else if (state->special_flags & JSONSL_SPECIALf_NULL ||
                            state->special_flags & JSONSL_SPECIALf_NAN) {
                    /* previous char was "n", are we parsing null or nan? */
-                   if (CUR_CHAR != 'u') {
+                   bool not_u = CUR_CHAR != 'u';
+                   bool not_a = tolower (CUR_CHAR) != 'a';
+                   if (not_u) {
                       state->special_flags &= ~JSONSL_SPECIALf_NULL;
                    }
-
-                   if (tolower(CUR_CHAR) != 'a') {
+                   if (not_a) {
                       state->special_flags &= ~JSONSL_SPECIALf_NAN;
                    }
+                   if (not_u && not_a) {
+                      VERIFY_SPECIAL ("null", 4);
+                   }
```

Comment:
This fixes a bug in `NaN`/`null` parsing in `jsonsl`. Previously: Any byte with value `'n'` would then be expected to be followed by a `u` or an `A`/`a` character, but this check did not error if `'n'` was followed by anything else, and the contiguous sequence of characters would be thrown away and ignored. That is, the string `[nope]` would parse as an empty array, because the `nope` would be erroneously ignored. This had fun effects as in `{"foo": nope}`, where we generate a key with no associated value. Woops.

Reply:
Great catch. This may be worth submitting a PR to jsonsl.

Reply:
I considered that, but I also discovered Roberto's existing open PR that has been sitting open for a good while 🙂.